common/journald/src/journalctl/mod.rs

```diff
@@ -0,0 +1,392 @@
+
+use std::convert::TryInto;
+use std::process::Stdio;
+
+const JOURNALCTL_CMD: &str = "journalctl";
+const KEY_MESSAGE: &str = "MESSAGE";
+const KEY_SYSTEMD_UNIT: &str = "_SYSTEMD_UNIT";
+const KEY_SYSLOG_IDENTIFIER: &str = "SYSLOG_IDENTIFIER";
+const KEY_CONTAINER_NAME: &str = "CONTAINER_NAME";
+const DEFAULT_APP: &str = "UNKNOWN_SYSTEMD_APP";
+
+pub struct JournaldExportDecoder {
+    state: AnyPartialState,
+}
+
+impl Default for JournaldExportDecoder {
+    fn default() -> Self {
+        JournaldExportDecoder {
+            state: Default::default(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum FieldValue {
+    Bytes(Vec<u8>),
+    Utf8(String),
+}
+
+impl FieldValue {
+    pub fn to_string_lossy(&self) -> String {
+        match self {
+            FieldValue::Bytes(b) => String::from_utf8_lossy(&b).to_string(),
+            FieldValue::Utf8(s) => s.clone(),
+        }
+    }
+}
+
+type JournalRecord = std::collections::HashMap<String, FieldValue>;
+
+// The actual parser for the journald export format
+fn decode_parser<'a, Input>(
+) -> impl Parser<Input, Output = JournalRecord, PartialState = AnyPartialState> + 'a
+where
+    Input: RangeStream<Token = u8, Range = &'a [u8]> + 'a,
+    // Necessary due to rust-lang/rust#24159
+    Input::Error: ParseError<Input::Token, Input::Range, Input::Position>,
+{
+    /*
+    Two journal entries that follow each other are separated by a double newline.
+
+     Journal fields consisting only of valid non-control UTF-8 codepoints are serialized as they are:
+         field name, followed by '=', followed by field data, followed by a newline as separator to the next field.
+     Other journal fields are serialized in a special binary safe way:
+         field name, followed by newline, followed by a binary 64bit little endian size value, followed by the binary field data,
+         followed by a newline as separator to the next field.
+     Entry metadata that is not actually a field is serialized like it was a field, but beginning with two underscores.
+     More specifically, __CURSOR=, __REALTIME_TIMESTAMP=, __MONOTONIC_TIMESTAMP= are introduced this way.
+     The order in which fields appear in an entry is undefined and might be different for each entry that is serialized.
+
+     And that's it.
+     */
+
+    any_partial_state(
+        (
+            many1::<std::collections::HashMap<_, _>, _, _>(
+                (
+                    // Get the key
+                    recognize(skip_many1(choice((letter(), token(b'_'))))).and_then(
+                        |bytes: &[u8]| {
+                            std::str::from_utf8(bytes)
+                                .map(|s| s.to_string())
+                                .map_err(StreamErrorFor::<Input>::other)
+                        },
+                    ),
+                    // Get the Value
+                    one_of(b"=\n"[..].iter().copied()).then_partial(|&mut c| {
```

Comment:
Pretty elegant, using `combine` is a good call.

Reply:
The original manual version wasn't too bad until I started trying to support partial reads, then it got very big, ugly and hard to follow.

Combine has it's own awkwardness, the lifetimes held me up for a few hours until I worked out how partial state works, but I agree, I still think it's worth it for the elegance.